parse.ml

```diff
@@ -9,9 +9,44 @@ let split_annotations program =
   let instructions = Array.map snd program in
   (instructions, annotations)
 
+exception Parse_error of string option * Lexing.position * Lexing.position
+
+let parse lexbuf =
+  (* see the Menhir manual for the description of
+     error messages support *)
+  let open MenhirLib.General in
+  let module Interp = Parser.MenhirInterpreter in
+  let input = Interp.lexer_lexbuf_to_supplier Lexer.token lexbuf in
+  let success prog = split_annotations prog in
+  let failure error_state =
+    let env = match error_state with
+      | Interp.HandlingError env -> env
+      | _ -> assert false in
+    match Interp.stack env with
+    | lazy Nil -> assert false
+    | lazy (Cons (Interp.Element (state, _, start_pos, end_pos), _)) ->
+      let message =
+        try Some (Parser_messages.message (Interp.number state)) with
+        | Not_found -> None in
+      raise (Parse_error (message, start_pos, end_pos))
+  in
+  Interp.loop_handle success failure input
+    (Parser.Incremental.program lexbuf.Lexing.lex_curr_p)
+
 let parse_string str =
   let lexbuf = Lexing.from_string str in
-  split_annotations (Parser.program Lexer.token lexbuf)
+  parse lexbuf
```

Comment:
Indeed, I can work on that. `parse_string` already uses the same auxiliary function. The reason why I haven't shared the error-reporting between `parse_file` and `parse_string` is that I first tried that, and that the `exit <error-code>` things break the testsuite (they stop the tests instead of continuing with other tests) while raising an uncaught exception does not. So we need something a bit more subtle, where the error-message-printing is done in both cases, but only one of them aborts. (Or maybe none, and then the users of `parse_file` with also die with an uncaught exception, which is maybe just fine.)